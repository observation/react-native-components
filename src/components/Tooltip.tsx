import React from 'react'
import { StyleSheet, View, Text, TouchableOpacity, ViewStyle, StyleProp } from 'react-native'

import { Icon, IconClose } from './Icon'
import LargeButton, { LargeButtonProps } from '../components/LargeButton'
import { IconName } from '../lib/Icons'
import textStyle from '../styles/text'
import theme from '../styles/theme'

type TooltipProps = {
  title: string
  text: string
  iconName?: IconName
  closable?: boolean
  onClose?: () => void
  buttons?: LargeButtonProps[]
  style?: StyleProp<ViewStyle>
  children?: React.ReactNode
  testID?: string
}

const Tooltip = ({
  iconName,
  title,
  text,
  closable = true,
  onClose,
  buttons = [],
  style,
  testID,
  children,
}: TooltipProps) => (
  <View style={[theme.shadow.ios, style]} testID={testID}>
    <View style={styles.tooltipContainer}>
      <View style={styles.tooltip}>
        <View style={{ flexDirection: 'row', alignItems: 'flex-start' }}>
          {iconName && (
            <View style={{ ...styles.iconContainer, marginRight: theme.margin.common }}>
              <Icon name={iconName} size={theme.icon.size.extraLarge} />
            </View>
          )}
          <View style={{ flex: 1 }}>
            <Text style={textStyle.lead}>{title}</Text>
          </View>
          {closable && (
            <View style={{ ...styles.iconContainer, marginLeft: theme.margin.common }}>
              <TouchableOpacity testID="close" onPress={onClose} activeOpacity={0.5} accessibilityLabel="close">
                <IconClose />
              </TouchableOpacity>
            </View>
          )}
        </View>
        {children}
        <View style={{ marginTop: theme.margin.half }}>
          <Text style={textStyle.body}>{text}</Text>
        </View>
        {buttons && buttons.length > 0 && (
          <View
            style={{
              marginTop: theme.margin.common,
              marginHorizontal: -theme.margin.half,
              flexDirection: 'row',
              alignItems: 'flex-start',
            }}
          >
            {buttons.map((button) => (
              <LargeButton
                disabled={button.disabled}
                secondary={button.secondary}
                title={button.title}
                key={button.title}
                style={{ margin: 0, marginHorizontal: theme.margin.half }}
                onPress={button.onPress}
              />
            ))}
          </View>
        )}
      </View>
    </View>
  </View>
)

export default Tooltip

export type { TooltipProps }

const styles = StyleSheet.create({
  tooltipContainer: {
    borderBottomLeftRadius: 16,
    borderBottomRightRadius: 16,
    overflow: 'hidden',
    backgroundColor: 'white',
    ...theme.shadow.android,
  },
  tooltip: {
    flexDirection: 'column',
    margin: theme.margin.common,
  },
  iconContainer: {
    height: theme.font.lineHeight.medium,
    justifyContent: 'center',
  },
})
