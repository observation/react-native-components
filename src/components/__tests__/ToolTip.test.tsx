import React from 'react'

import { render, fireEvent } from '@testing-library/react-native'

import Tooltip from '../Tooltip'

const onCloseTooltip = jest.fn()

describe('Tooltip', () => {
  describe('Rendering', () => {
    test('With icon', () => {
      // GIVEN
      const { toJSON } = render(<Tooltip icon={{ name: 'info-circle' }} title="Title" text="Text" />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('Without icon', () => {
      // GIVEN
      const { toJSON } = render(<Tooltip title="Title" text="Text" />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('Without icon or close button', () => {
      // GIVEN
      const { toJSON } = render(<Tooltip title="Title" text="Text" closable={false} />)
      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With button', () => {
      // GIVEN
      const { toJSON } = render(
        <Tooltip title="Title" text="Text" closable={false} buttons={[{ onPress: () => {}, title: 'button title' }]} />,
      )
      // THEN
      expect(toJSON()).toMatchSnapshot()
    })
  })

  describe('Interaction', () => {
    test('Click on close', () => {
      // GIVEN
      const { getByTestId } = render(
        <Tooltip title="Title" text="Text" onClose={onCloseTooltip} icon={{ name: 'info-circle' }} />,
      )
      // WHEN
      fireEvent.press(getByTestId('close'))

      // THEN
      expect(onCloseTooltip).toBeCalled()
    })

    test('Click on a button', () => {
      // GIVEN
      const onPress = jest.fn()
      const { getByText } = render(
        <Tooltip title="Title" text="Text" buttons={[{ onPress, title: 'Action!' }]} icon={{ name: 'info-circle' }} />,
      )

      // WHEN
      fireEvent.press(getByText('Action!'))

      // THEN
      expect(onPress).toBeCalled()
    })
  })
})
