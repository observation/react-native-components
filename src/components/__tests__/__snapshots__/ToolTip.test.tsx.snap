// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Rendering With button 1`] = `
<View
  style={
    [
      {
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "elevation": 6,
        "overflow": "hidden",
        "shadowColor": "#939393",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#212121",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 20,
              }
            }
          >
            Title
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 20,
            }
          }
        >
          Text
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "marginHorizontal": -8,
            "marginTop": 16,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#0066B1",
              "borderRadius": 4,
              "height": 32,
              "justifyContent": "center",
              "margin": 0,
              "marginHorizontal": 8,
              "opacity": 1,
              "overflow": "hidden",
            }
          }
          testID="touchable-opacity"
        >
          <View
            style={
              {
                "flexDirection": "row",
                "marginHorizontal": 16,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#212121",
                    "fontFamily": "Ubuntu",
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "bold",
                    "lineHeight": 20,
                    "textAlignVertical": "center",
                  },
                  {
                    "color": "#FFFFFF",
                  },
                ]
              }
            >
              button title
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Tooltip Rendering With icon 1`] = `
<View
  style={
    [
      {
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "elevation": 6,
        "overflow": "hidden",
        "shadowColor": "#939393",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "height": 20,
              "justifyContent": "center",
              "marginRight": 16,
            }
          }
        >
          <Icon
            color="#0066B1"
            name="circle-info"
            size={18}
            type="light"
          />
        </View>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#212121",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 20,
              }
            }
          >
            Title
          </Text>
        </View>
        <View
          style={
            {
              "height": 20,
              "justifyContent": "center",
              "marginLeft": 16,
            }
          }
        >
          <View
            accessibilityLabel="close"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="close"
          >
            <Icon
              color="#939393"
              name="xmark"
              size={18}
              type="light"
            />
          </View>
        </View>
      </View>
      <View
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 20,
            }
          }
        >
          Text
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Tooltip Rendering Without icon 1`] = `
<View
  style={
    [
      {
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "elevation": 6,
        "overflow": "hidden",
        "shadowColor": "#939393",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#212121",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 20,
              }
            }
          >
            Title
          </Text>
        </View>
        <View
          style={
            {
              "height": 20,
              "justifyContent": "center",
              "marginLeft": 16,
            }
          }
        >
          <View
            accessibilityLabel="close"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="close"
          >
            <Icon
              color="#939393"
              name="xmark"
              size={18}
              type="light"
            />
          </View>
        </View>
      </View>
      <View
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 20,
            }
          }
        >
          Text
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Tooltip Rendering Without icon or close button 1`] = `
<View
  style={
    [
      {
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "elevation": 6,
        "overflow": "hidden",
        "shadowColor": "#939393",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#212121",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 20,
              }
            }
          >
            Title
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 20,
            }
          }
        >
          Text
        </Text>
      </View>
    </View>
  </View>
</View>
`;
